New py of making an order

Richymaestro opened the pull request:
Made a new version of the order.py, in style of OOP Python. Let me know if it works and if you have remarks!

Richymaestro committed:
New py of making an order

bh2smith commented on src/CowOrder.py:
```diff
@@ -0,0 +1,92 @@
+        self.feeAddress = feeAddress
+    
+    def setFee(self):
+        requestFee = requests.get(base_url + feeAddress + sellToken + "&buyToken=" + buyToken + "&amount=" +sellAmount + "&kind=" + kind)
```
Generally speaking, I am not a fan of raw string concatenation. I think the code would be much more readable and reliable if we were to use python's string formatting with variable names as follow

```py
url = "{baseUrl}fee?sellToken={sellToken}&buyToken={buyToken}&amount={amount}&kind={kind}".format(
    baseUrl=base_url, 
    sellToken=self.sell_token, 
    buyToken=self.buy_token,
    amount=self.sell_amount,
    kind=self.kind
)
```

Richymaestro replied:
Yes, I agree with that, will change this one

bh2smith commented on src/CowOrder.py:
```diff
@@ -0,0 +1,92 @@
+        jsonorder = json.dumps(self, default=lambda obj: obj.__dict__, indent=4)
+        self.orders = orders
+        self.base_url = base_url
+        return requests.post(self.base_url + self.orders + jsonorder)
```
If we look at the post made [here]() in the original script 

```
r = requests.post(base_url + "orders", json=j)
```
this is a string concatenation of the base url with the string "orders" followed by the json data of the order

```suggestion
        return requests.post("{base_url}orders".format(base_url), json=jsonorder)
```

bh2smith commented on src/CowOrder.py:
```diff
@@ -0,0 +1,92 @@
+        self.buyTokenBalance: String() = buyTokenBalance 
+    
+    def toJson(self):
+        jsonorder = json.dumps(self, default=lambda obj: obj.__dict__, indent=4)
```
This method is called "toJson" so we would expect that it merely returns the jsonorder declared in the first line here. However, this method appears to be posting an order.

I would suggest 
```suggestion
    def to_json(self):
        return json.dumps(self, default=lambda obj: obj.__dict__, indent=4)
```

followed by another method `post_order` which does something like
```py
return requests.post("{base_url}orders".format(base_url), json=self.to_json())
```

bh2smith commented on src/CowOrder.py:
```diff
@@ -0,0 +1,92 @@
+        self.appData: Bytes(32) = appData
+        self.feeAmount: Uint(256) = str(Fee.setFee(self))
+        self.partiallyFillable: Boolean() = partiallyFillable
+        self.signature = Signature.getSig(self)
```
While this is a pretty good idea we may also want to update certain parameters of an order (like the fee amount) if it changes. In this case I might also include a method `update_fee` which fetches a new fee quote and updates the value. Along with this, then we want a method like "Order.sign()" which fills the signature filed based on all the current values.

bh2smith reviewed (changes requested):
Hello @Richymaestro and thanks for your very nice looking contribution! Very glad to have you interested in the project.

Everything is looking very good so far, but there are a few places here which aren't quite correct and others that could be improved.

Given that this is quite involved (i.e. contains the introduction of several classes and methods). I might also suggest introducing them each as separate PRs. It is, however pretty clear that all the classes are quite interconnected, so we could allow them for now pending some cleanup and unit testing afterwards.

Each class should be in its own file
```
fee.py
signature.py
order.py
```

Another thing that would probably be more appropriate is the include your changes directly in the existing file. Once you can actually manage to post an order with the content of this code, it would be really good if you could include a "Test Plan" in the PR description demonstrating that an order was placed using it.

Richymaestro committed:
Update src/CowOrder.py

Co-authored-by: Benjamin Smith <bh2smith@users.noreply.github.com>

Richymaestro commented:
Thanks for your comments, I will add them tomorrow!

Richymaestro committed:
Update CowOrder.py

URL to request fee in more pythonic string format

Richymaestro committed:
Update CowOrder.py

Made some changes, only problem now is that the JSON also includes base_url. Do you have a suggestions how to work around this?

bh2smith commented on src/CowOrder.py:
```diff
@@ -0,0 +1,101 @@
+        self.sellAmount = sellAmount
+        self.kind = kind
+    
+    def setFee(self):
```
This method would be more appropriately named as get_fee since it fetches it and returns it.

Richymaestro commented:
Made a new PR with updated code, will close this one
